eth/db/header.py

```diff
@@ -296,6 +287,47 @@ def _persist_header_chain(
 
         return tuple(), tuple()
 
+    @classmethod
+    def _handle_gap_change(cls,
+                           db: DatabaseAPI,
+                           gap_change: GapChange,
+                           header: BlockHeaderAPI,
+                           genesis_parent_hash: Hash32) -> None:
+
+        if gap_change not in GAP_WRITES:
+            return
+
+        if gap_change is GapChange.GapFill:
+            expected_child = cls._get_canonical_head(db)
```

Comment:
Canonical head as the child of the gap seems wrong to me. What about this series of added headers:

```
A2 (set checkpoint)
A3
B0:2
A0:2
```
(using a and b from the test, and python index/slice rules)

Reply:
You are absolutely right (and I admire how you spot these things during a review!). I added a fix + test case to cover this.